cmd/agent/daemon/state/events_pipeline.go

```diff
@@ -19,125 +20,278 @@ func (c *Controller) runEventsPipeline(ctx context.Context) error {
 	c.log.Info("running events pipeline")
 	defer c.log.Info("events pipeline done")
 
+	var currentEventsCount int
+	lastFlushedAt := c.nowFunc()
+	groups := map[uint64]*castpb.ContainerEvents{}
+
+	ticker := time.NewTicker(c.cfg.EventsFlushInterval)
+	defer ticker.Stop()
+
+	send := func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		batch := &castpb.ContainerEventsBatch{}
+		for _, group := range groups {
+			if len(group.Items) > 0 {
+				batch.Items = append(batch.Items, group)
+			}
+		}
+		if len(batch.Items) == 0 {
+			return
+		}
+
+		c.log.Debugf("sending events batch, events=%d, groups=%d", currentEventsCount, len(batch.Items))
+		if err := c.exporters.ContainerEventsSender.Send(ctx, batch); err != nil {
+			c.log.Errorf("sending events batch: %s", err)
+		}
+
+		// Reset state after data is sent.
+		for _, group := range groups {
+			group.Items = group.Items[:0]
+		}
+		lastFlushedAt = c.nowFunc()
+		currentEventsCount = 0
+	}
+
+	handleEvent := func(e *ebpftypes.Event) {
+		group, found := groups[e.Context.CgroupID]
+		if !found {
+			group = c.newContainerEventsGroup(e)
+			groups[e.Context.CgroupID] = group
+		}
+		pbEvent := &castpb.ContainerEvent{}
+		c.fillProtoContainerEvent(pbEvent, e, nil)
+
+		if c.enrichmentService.Enqueue(&enrichment.EnrichedContainerEvent{
+			Event:     pbEvent,
+			EbpfEvent: e,
+		}) {
+			return
+		}
+
+		group.Items = append(group.Items, pbEvent)
+		currentEventsCount++
+	}
+
+	handleSignatureEvent := func(e *ebpftypes.Event, signatureEvent *castpb.SignatureEvent) {
+		group, found := groups[e.Context.CgroupID]
+		if !found {
+			group = c.newContainerEventsGroup(e)
+			groups[e.Context.CgroupID] = group
+		}
+		pbEvent := &castpb.ContainerEvent{}
+		c.fillProtoContainerEvent(pbEvent, e, signatureEvent)
+
+		group.Items = append(group.Items, pbEvent)
+		currentEventsCount++
+	}
+
+	handleEnrichedEvent := func(e *enrichment.EnrichedContainerEvent) {
+		group, found := groups[e.EbpfEvent.Context.CgroupID]
+		if !found {
+			// If enriched event is not found here most likely container was removed.
+			return
+		}
+
+		group.Items = append(group.Items, e.Event)
+		currentEventsCount++
+	}
+
+	// Since we have multiple channels with different priority reading from the channel is split into
+	// multiple non-blocking selects. Except the last one. It's important to keep the last select blocking
+	// otherwise it will consume CPU resources.
 	for {
+		// Top priority, handle context cancel and flush remaining data.
 		select {
 		case <-ctx.Done():
+			send()
 			return ctx.Err()
-		case e := <-c.tracer.Events():
-			pbEvent := c.toProtoEvent(e)
-			if c.enrichmentService.Enqueue(&enrichment.EnrichRequest{
-				Event:     pbEvent,
-				EbpfEvent: e,
-			}) {
+		default:
+		}
+
+		// High priority, handle deleted first containers.
```

Comment:
Why do we need the split here? It should not matter if we first remove events or flush all of them.

Reply:
Since select picks any next queue randomly and if you have high events rate, container deletion events could take quite some time to actually be picked by select. But we need to process it as soon as possible. CgroupID could be reused by a new container. In general probably need to rethink the key here and do not use cgroupID as a key. Maybe hashing containerid to digest would be safer. But not sure if it's worth reworking it now.

Reply:
It would be nice if you could add this context as a comment, as it is not straight forward to understand.